### PR TITLE
Support for np.flatnonzero

### PR DESCRIPTION
**Pull request by rjenc29**
Support for `np.flatnonzero`

Initial commit for CI.

### CODE REVIEW

**Comment by rjenc29** [comment]
Hi

This is a pretty small PR - not much to flag up other than Numpy behaviour changed in 1.15 w.r.t. handling of array-likes: I have broadly followed suit, with a judicious if-gate and TypingError.  

If you get a change to review, I'd appreciate your feedback and will follow up as soon as I can.

Cheers

File: numba/targets/arraymath.py [hunk 1]
```diff
@@ -2396,6 +2396,21 @@ def np_corrcoef_impl_single_variable(x, y=None, rowvar=True):
...
+@overload(np.flatnonzero)
+def np_flatnonzero(a):
+
+    if numpy_version < (1, 15):
+        if not isinstance(a, types.Array):
+            raise TypingError("Argument 'a' must be an array")
+            # numpy raises an Attribute error with:
+            # 'xxx' object has no attribute 'ravel'
+
+    def impl(a):
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: I think that this function needs guarding with the **new** `type_can_asarray` helper:

https://github.com/numba/numba/blob/8bfe1e845d09b7ecc2055a2be4f90a066145bab6/numba/numpy_support.py#L599-L604

so e.g. someone randomly passing a string to this function doesn't get a weird error message from `np.asarray` about whatever it doesn't like about a string.

Comment (rjenc29) [reply numba/targets/arraymath.py]: Good call.

I added a couple of lines to handle invalid array-likes; all seems fine except for the python 2.7 builds, which error with:  ```TypeError: can't unbox str type```

Let me know what the best course of action is and I will tweak accordingly.

Comment (stuartarchibald) [reply numba/targets/arraymath.py]: Thanks. String support isn't present in Numba for Python 2.7, which is probably the root cause, think it's fine to add a guard around Python version in the test inputs for this specific case.

**Comment by rjenc29** [comment]
Excepting that Travis has given himself the weekend off, I think the last set of tweaks should get this MR into shape / worth a re-review whenever you get a mo.

Cheers

File: numba/tests/test_array_manipulation.py [hunk 2]
```diff
@@ -740,6 +746,78 @@ def check(x):
...
+        yield np.nan
+        yield 0
+        yield 1
+        yield False
+        yield (True, False, True)
+        yield 2 + 1j
+        # the following are not array-like, but numpy 1.15+ does not raise
+        yield None
+        # string support isn't present in Numba for Python 2.7
+        if sys.version_info >= (3,):
```
Comment (stuartarchibald) [inline numba/tests/test_array_manipulation.py]: Think 3.4 was minimum for unicode support so this condition probably ought to read `if not sys.version_info < (3,)`

File: numba/targets/arraymath.py [hunk 3]
```diff
@@ -2396,6 +2396,34 @@ def np_corrcoef_impl_single_variable(x, y=None, rowvar=True):
...
+            raise TypingError("Argument 'a' must be an array")
+            # numpy raises an Attribute error with:
+            # 'xxx' object has no attribute 'ravel'
+
+    if type_can_asarray(a):
+        def impl(a):
+            arr = np.asarray(a)
+            return np.nonzero(np.ravel(arr))[0]
+    else:
+        if utils.MACHINE_BITS == 32:
```
Comment (stuartarchibald) [inline numba/targets/arraymath.py]: Think this branching resolves to `dtype=numba.types.intp` ?

**Comment by stuartarchibald** [comment]
Thanks for the fix-ups, looks good to me, can be marked for merge on CI pass.

**Pull request merged**